### PR TITLE
Add cache_prefix setting (fixes #227)

### LINKED ISSUE #227: Add key prefix option for cache
* Add a setting ``cliquet.cache_prefix`` (default: ``''``)
* Every cache entry key is prefixed by ``self.prefix`` in cache backend classes

### PR DESCRIPTION
**Pull request by lavish205**
issue #227

### CODE REVIEW

**Comment by lavish205** [comment]
@leplatrem I'm getting `cache_prefix` KeyError as `config.get_settings()` is not returning settings defined in `DEFAULT_SETTINGS`

**Comment by leplatrem** [comment]
It looks perfectly correct.

But when test are ran:
```
$ nosetests -s cliquet.tests.test_cache
```

I also get this error: 
```
Traceback (most recent call last):
  File "/home/mathieu/Code/Mozilla/cliquet/cliquet/tests/test_cache.py", line 40, in setUp
    self.cache = self.backend.load_from_config(self._get_config())
  File "/home/mathieu/Code/Mozilla/cliquet/cliquet/cache/memory.py", line 55, in load_from_config
    prefix = settings['cache_prefix']
KeyError: 'cache_prefix'
```

That's because the tests setup do not use DEFAULT_SETTINGS.
See, you should add something like this on `MemoryCacheTest` https://github.com/mozilla-services/cliquet/blob/master/cliquet/tests/test_cache.py#L162-L165

**Comment by lavish205** [comment]
@leplatrem I have fixed this KeyError by `.get()` method, hope this works

**Comment by lavish205** [comment]
#227

**Comment by leplatrem** [comment]
> hope this works.

Nope actually. You should leave ``['cache_prefix']`` and set its value in the tests like I pointed out.

Here is a small list of things to do before this could be merged:

* [ ] Generalize the change to other backends (redis and postgresql)
* [ ] Add some tests in `test_cache.py` to ensure that the prefix value is used to fetch/delete the data associated to a key
* [ ] Add a test to ensure that the default prefix value is empty
* [ ] Update configuration documentation
* [ ] Update changelog
* [ ] Add yourself to contributors

**Comment by leplatrem** [comment]
@lavish205 you could do something like that (I didn't test it):
```python
def test_get_uses_prefix(self):
    settings_prefix = self.settings.copy()
    settings_prefix['cache_prefix'] = 'prefix_'
    config_prefix = self._get_config(settings=settings_prefix)
    
    # Instantiate a cache backend with a prefix:
    backend_prefix = self.backend.load_from_config(config_prefix)

    # Set a value with a cache that has no prefix:
    self.cache.set('prefix_key', 'foo')

    # Obtain the value with the cache that has a prefix:
    obtained = backend_prefix.get('key')
    self.assertEqual(obtained, 'foo')

```

**Comment by lavish205** [comment]
@leplatrem I think something is wrong with the code as test cases are not working as expected.

**Comment by leplatrem** [comment]
@lavish205 yes, first you shouldn't set the prefix to `123`, so both `self.cache` and `backend_prefix` have a prefix.

Then, in order to make tests pass, you should use the `self.prefix` in `def get()` in the `Cache` backend class.

**Comment by lavish205** [comment]
@leplatrem value of `backend_prefix` is `{'prefix': 'prefix_', '_store': {'prefix_key': 'bar'}, '_ttl': {}`, so can't access `prefix_key` on root of dict.

File: cliquet/tests/test_cache.py [hunk 1]
```diff
@@ -156,12 +159,28 @@ def test_ping_returns_false_if_unavailable(self):
...
+        config_prefix = self._get_config(settings=settings_prefix)
+
+        # initiating cache backend with prefix:
+        backend_prefix = self.backend.load_from_config(config_prefix)
+
+        # set a value with a cache that has no prefix
+        backend_prefix.set('prefix_key', 'foo')
+
+        # obtain the value of cache that has prefix
+        obtained = backend_prefix.get('prefix_key')
```
Comment (leplatrem) [inline cliquet/tests/test_cache.py]: This should be `get('key')` since this backend has prefix `prefix_` !

**Comment by lavish205** [comment]
@leplatrem can you please check, if there is need to add any test case?

File: cliquet/cache/memory.py [hunk 2]
```diff
@@ -25,20 +26,24 @@ def flush(self):
         self._store = {}
 
     def ttl(self, key):
+        key = self.prefix + key
```
Comment (Natim) [inline cliquet/cache/memory.py]: I would rather use format here: ``key = "{}{}".format(self.prefix, key)`` or ``key = "%s%s" % (self.prefix, key)``

File: cliquet/cache/memory.py [hunk 3]
```diff
@@ -25,20 +26,24 @@ def flush(self):
...
 
     def expire(self, key, ttl):
+        key = self.prefix + key
         self._ttl[key] = utils.msec_time() + int(ttl * 1000.0)
 
     def set(self, key, value, ttl=None):
-        self._store[key] = value
         if ttl is not None:
             self.expire(key, ttl)
+        key = self.prefix + key
```
Comment (Natim) [inline cliquet/cache/memory.py]: You need to do that before calling expire because you should expire the key with the right prefix too,

File: cliquet/cache/memory.py [hunk 4]
```diff
@@ -46,9 +51,13 @@ def get(self, key):
...
         self._ttl.pop(key, None)
         self._store.pop(key, None)
 
 
 def load_from_config(config):
-    return Cache()
+    settings = config.get_settings()
+    prefix = settings['cache_prefix']
+    backend = Cache(prefix=prefix)
+    return backend
```
Comment (Natim) [inline cliquet/cache/memory.py]: You can probably do ``return Cache(prefix=settings['cache_prefix'])`` and save some intermediary variables.

File: cliquet/tests/test_cache.py [hunk 5]
```diff
@@ -156,12 +159,86 @@ def test_ping_returns_false_if_unavailable(self):
     def test_ping_logs_error_if_unavailable(self):
         pass
 
+    def test_cache_prefix_is_set(self):
+        settings_prefix = self.settings.copy()
+        settings_prefix['cache_prefix'] = 'prefix_'
+        config_prefix = self._get_config(settings=settings_prefix)
+
+        # initiating cache backend with prefix:
+        backend_prefix = self.backend.load_from_config(config_prefix)
```
Comment (Natim) [inline cliquet/tests/test_cache.py]: Why cannot you just do ``backend_prefix = self.backend.load_from_config(prefix="prefix_")`` here?

File: cliquet/tests/test_cache.py [hunk 6]
```diff
@@ -156,12 +159,86 @@ def test_ping_returns_false_if_unavailable(self):
...
+
+        # set a value with a cache that has no prefix
+        backend_prefix.set('key', 'foo')
+
+        # obtain the value of cache that has prefix
+        obtained = backend_prefix.get('key')
+
+        self.assertEqual(obtained, 'foo')
```
Comment (Natim) [inline cliquet/tests/test_cache.py]: Where do you check that redis was called with ``prefix_key``?

File: cliquet/tests/test_cache.py [hunk 7]
```diff
@@ -156,12 +159,86 @@ def test_ping_returns_false_if_unavailable(self):
...
+        config_prefix = self._get_config(settings=settings_prefix)
+
+        # initiating cache backend with prefix:
+        backend_prefix = self.backend.load_from_config(config_prefix)
+
+        # set a value with a cache that has no prefix
+        backend_prefix.set('key', 'foo')
+
+        # deleting data
+        backend_prefix.delete('key')
```
Comment (Natim) [inline cliquet/tests/test_cache.py]: You should try to get the key and make sure it has been deleted.

File: cliquet/tests/test_cache.py [hunk 8]
```diff
@@ -156,12 +159,86 @@ def test_ping_returns_false_if_unavailable(self):
...
+
+    def test_prefix_value_used_with_ttl(self):
+        settings_prefix = self.settings.copy()
```
Comment (Natim) [inline cliquet/tests/test_cache.py]: Since you are doing this code in all the test cases it probably worth to put in a ``def setUp(self)`` method.

File: cliquet/tests/test_cache.py [hunk 9]
```diff
@@ -156,12 +159,86 @@ def test_ping_returns_false_if_unavailable(self):
...
+        settings_prefix['cache_prefix'] = 'prefix_'
+        config_prefix = self._get_config(settings=settings_prefix)
+
+        # initiating cache backend with prefix:
+        backend_prefix = self.backend.load_from_config(config_prefix)
+
+        # set a value with a cache that has no prefix
+        backend_prefix.set('key', 'foo', 10.0)
+        obtained = backend_prefix.ttl('key')
```
Comment (Natim) [inline cliquet/tests/test_cache.py]: Same here, we should validate that redis is always called with the right key.

**Comment by Natim** [comment]
That's a great job @lavish205 :+1: Thank you for working on it :)

File: cliquet/tests/test_cache.py (hunk 5 above)
Comment (leplatrem) [inline cliquet/tests/test_cache.py]: Because `load_from_config()` takes a config object :)

File: cliquet/tests/test_cache.py (hunk 8 above)
Comment (leplatrem) [inline cliquet/tests/test_cache.py]: Or add a helper `def get_backend_prefix()`

File: cliquet/cache/memory.py (hunk 2 above)
Comment (leplatrem) [inline cliquet/cache/memory.py]: Why ?

File: cliquet/cache/memory.py (hunk 3 above)
Comment (lavish205) [inline cliquet/cache/memory.py]: @Natim We can't do that because in `def expire()` we are again modifying the key. If we do it before calling the `self.expire()` then `key` will be modified to `prefix_prefix_key`

File: cliquet/tests/test_cache.py (hunk 6 above)
Comment (lavish205) [inline cliquet/tests/test_cache.py]: I didn't get you.

File: cliquet/cache/memory.py (hunk 3 above)
Comment (Natim) [inline cliquet/cache/memory.py]: Good point :+1:

File: cliquet/tests/test_cache.py (hunk 9 above)
Comment (lavish205) [inline cliquet/tests/test_cache.py]: @Natim can you tell me how to validate that redis is always called with the right key?

File: cliquet/tests/test_cache.py [hunk 10]
```diff
@@ -15,7 +15,7 @@
 
 class CacheBaseTest(unittest.TestCase):
     def setUp(self):
-        self.cache = CacheBase()
+        self.cache = CacheBase(cache_prefix='')
```
Comment (leplatrem) [inline cliquet/tests/test_cache.py]: Should be `prefix=''` here

File: cliquet/tests/test_cache.py (hunk 6 above)
Comment (leplatrem) [inline cliquet/tests/test_cache.py]: The idea is to use two instances, one with prefix and the other one without.

Comment (Natim) [inline cliquet/tests/test_cache.py]: This idea is really good, but doesn't work with the MemoryBackend because they are not sharing the same store.

**Comment by lavish205** [comment]
@leplatrem what to update in configuration documentation?

**Comment by Natim** [comment]
@lavish205 I believe it is just about adding the ``cache_prefix`` configuration settings as you told me on IRC.

> 13:52:43 <lavish_> do I need to add prefix changes here?
> http://kinto.readthedocs.org/en/latest/configuration/settings.html#cache

**Comment by lavish205** [comment]
@Natim what exactly I have write under "What does it do?"  column ?

**Comment by Natim** [comment]
@lavish205 Something like: ``It add a prefix to cache key names so that multiple kinto server can use the same cache database without disturbing each other.``

File: CHANGELOG.rst [hunk 11]
```diff
@@ -10,6 +10,7 @@ This document describes changes between each past release.
 **New features**
 
 - Default console log renderer now has colours (#671)
+- ``Prefix`` key option added for cache.
```
Comment (leplatrem) [inline CHANGELOG.rst]: The setting is called `cache_prefix` no?

**Comment by Natim** [comment]
I have added some documentation and a test to make sure ``cache.get`` added the prefix.

Thank you @lavish205 for your work on this.

**Comment by leplatrem** [comment]
Congrats @lavish205 for keeping the faith and going through all this!

Thank you for getting involved with patience !

**Comment by lavish205** [comment]
Thanks to both of you @leplatrem @Natim :)
